pkg/clusteragent/externalmetrics/provider.go

```diff
@@ -0,0 +1,152 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+// +build kubeapiserver
+
+package externalmetrics
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/kubernetes-incubator/custom-metrics-apiserver/pkg/provider"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver/common"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver/leaderelection"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/autoscalers"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const (
+	autogenExpirationPeriodHours int64 = 3
+)
+
+type datadogMetricProvider struct {
+	apiCl            *apiserver.APIClient
+	store            DatadogMetricsInternalStore
+	autogenNamespace string
+}
+
+func NewDatadogMetricProvider(ctx context.Context, apiCl *apiserver.APIClient) (provider.ExternalMetricsProvider, error) {
+	if apiCl == nil {
+		return nil, fmt.Errorf("Impossible to create DatadogMetricProvider without valid APIClient")
+	}
+
+	le, err := leaderelection.GetLeaderEngine()
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create DatadogMetricProvider as LeaderElection failed with: %v", err)
+	}
+
+	aggregator := config.Datadog.GetString("external_metrics.aggregator")
+	rollup := config.Datadog.GetInt("external_metrics_provider.rollup")
+	setQueryConfigValues(aggregator, rollup)
+
+	refreshPeriod := config.Datadog.GetInt64("external_metrics_provider.refresh_period")
+	retrieverMetricsMaxAge := int64(math.Max(config.Datadog.GetFloat64("external_metrics_provider.max_age"), float64(3*rollup)))
+	autogenNamespace := common.GetResourcesNamespace()
+
+	provider := &datadogMetricProvider{
+		apiCl:            apiCl,
+		store:            NewDatadogMetricsInternalStore(),
+		autogenNamespace: autogenNamespace,
+	}
+
+	// Start MetricsRetriever, only leader will do refresh metrics
+	dogCl, err := autoscalers.NewDatadogClient()
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create DatadogMetricProvider as DatadogClient failed with: %v", err)
+	}
+
+	metricsRetriever, err := NewMetricsRetriever(refreshPeriod, retrieverMetricsMaxAge, autoscalers.NewProcessor(dogCl), le.IsLeader, &provider.store)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create DatadogMetricProvider as MetricsRetriever failed with: %v", err)
+	}
+	go metricsRetriever.Run(ctx.Done())
+
+	// Start AutoscalerWatcher, only leader will flag DatadogMetrics as Active/Inactive
+	// WPAInformerFactory is nil when WPA is not used. AutoscalerWatcher will check value itself.
```

Comment:
(note to self) The autoscaler watcher only uses the wpa lister if the informer is not nil (thus it's enabled)
https://github.com/DataDog/datadog-agent/pull/5384/files#diff-c34c8104d441fefbe8af5ba433cbb98cR215